### PR TITLE
Remove isRootRelation knowledge from consumers

### PR DESCRIPTION
**Pull request by mfussenegger**
This commit changes the consumers contract.
A consumer now plans as distributed as possible and no longer cares if
it has a parent relation or not.

Any information that is necessary to merge this distributed result has
to be provided in a ResultDescription which can be picked up by a parent
to create a handler-merge.

This changes some execution plans slightly. For example, an insert with
a distributed group by in a subquery will now execute the inserts from
the reducer nodes instead of from the handler node.

### CODE REVIEW

**Comment by mfussenegger** [comment]
@seut & @matriv please review.

Probably makes sense to start with
https://github.com/crate/crate/pull/4352/files#diff-943ee35294202b4729b4ddafe9e4785b
and https://github.com/crate/crate/pull/4352/files#diff-bee159e9a13e581e2a43daf4a1df0274

File: sql/src/main/java/io/crate/planner/projection/builder/ProjectionBuilder.java [hunk 1]
```diff
@@ -115,9 +117,17 @@ public GroupProjection groupProjection(
...
+    public static FilterProjection filterProjection(Collection<? extends Symbol> inputs, QueryClause queryClause) {
         InputCreatingVisitor.Context context = new InputCreatingVisitor.Context(inputs);
-        query = inputVisitor.process(query, context);
+        Symbol query;
+        if (queryClause.hasQuery()) {
+            query = inputVisitor.process(queryClause.query(), context);
+        } else if (queryClause.noMatch()) {
+            query = Literal.BOOLEAN_FALSE;
+        } else {
+            // TODO: return null and skip FilterProjection?
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/projection/builder/ProjectionBuilder.java]: would leave it like that, the overhead is ignorable afaik

File: sql/src/main/java/io/crate/planner/Merge.java [hunk 2]
```diff
@@ -22,20 +22,144 @@
...
+        if (ExecutionPhases.executesOnHandler(plannerContext.handlerNode(), resultDescription.nodeIds())) {
+            for (Projection projection : projections) {
+                assert projection.outputs().size() == resultDescription.numOutputs()
+                    : "projection must not affect numOutputs";
+                subPlan.addProjection(projection, null, null, null, null);
+            }
+            if (topN != null) {
+                subPlan.addProjection(topN, TopN.NO_LIMIT, 0, resultDescription.numOutputs(), null);
+            }
+            // resultDescription.orderBy can be ignored here because it is only relevant to do a sorted merge
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/Merge.java]: the main factor for not looking at the order is that there is only one bucket ... pls mention

File: sql/src/main/java/io/crate/planner/Merge.java [hunk 3]
```diff
@@ -22,20 +22,144 @@
...
+            // (of a pre-sorted result)
+            return subPlan;
+        }
+        Collection<String> handlerNodeIds;
+        // ideally we would use something different as an indicator for direct-result or push/paging
+        if (subPlan instanceof Collect) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/Merge.java]: what about paging? can we do this check  at once?

File: sql/src/main/java/io/crate/planner/Merge.java [hunk 4]
```diff
@@ -22,20 +22,144 @@
...
+            }
+            if (topN != null) {
+                subPlan.addProjection(topN, TopN.NO_LIMIT, 0, resultDescription.numOutputs(), null);
+            }
+            // resultDescription.orderBy can be ignored here because it is only relevant to do a sorted merge
+            // (of a pre-sorted result)
+            return subPlan;
+        }
+        Collection<String> handlerNodeIds;
+        // ideally we would use something different as an indicator for direct-result or push/paging
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/Merge.java]: pls elaborate on the background why this is not generic

File: sql/src/main/java/io/crate/planner/ResultDescription.java [hunk 5]
```diff
@@ -34,4 +38,43 @@
      * If it's empty it is on the handler node.
      */
     Collection<String> nodeIds();
+
+    /**
+     * OrderBy information a parent can use to do a sorted-merge.
+     * This needs to be set if (and only if) the result is pre-sorted.
+     *
+     * This DOES NOT cause the parent relation to do a FULL sort.
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/ResultDescription.java]: s/cause/require/

File: sql/src/main/java/io/crate/planner/ResultDescription.java [hunk 6]
```diff
@@ -34,4 +38,43 @@
...
+     */
+    @Nullable PositionalOrderBy orderBy();
+
+    /**
+     * @return Limit that needs to be applied if the result is merged.
+     *         This does not indicate if a limit was applied to generate the result,
+     *
+     *         Therefore if a result was produced by applying the final limit it is allowed to return
+     *         {@link io.crate.operation.projectors.TopN#NO_LIMIT} to indicate that no additional limit needs to be applied.
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/ResultDescription.java]: hm, how does the consumer of this result know if it should page or not if the description might say no_limit

**Comment by dobe** [comment]
ftr: introduce ResultDescripiton:maxRowsPerNode

File: sql/src/main/java/io/crate/planner/Merge.java [hunk 7]
```diff
@@ -22,20 +22,165 @@
...
+            projections.add(topN);
+            return projections;
+        }
+    }
+
+    public static Plan mergeToHandler(Plan subPlan, Planner.Context plannerContext) {
+        return mergeToHandler(subPlan, plannerContext, Collections.emptyList());
+    }
+
+    public Merge(Plan subPlan,
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/Merge.java]: I would move the constructor to the top of the class

**Pull request merged**